V2 of embedded mode tutorial (with supporting `modes` guide) (#5771)
keystonejs/keystone: docs/pages/tutorials/embedded-mode-with-nextjs-frontend.mdx

import { Markdown } from '../../components/Page';

# How to embed Keystone in a NextJS app

Keystone now supports SQLite! 🎉

This makes it possible to run a local instance of Keystone from the same place you keep your frontend code and commit everything to Git. If you're happy to apply content changes in a local-only dev environment, and don't need a writeable API in production this use case may be handy for you.

---

> [Keystone Lite](https://github.com/keystonejs/keystone/tree/master/examples/next-lite) is a streamlined [mode](/guides/modes) that you can operate Keystone with. It takes advantage of SQLite’s embedded local storage capabilities to give you a powerful CMS that you can run locally and store content in the same place you keep your frontend code.

Author: It feels weird to me too 🤔 
But agree not a blocker 👍

The tutorial will show you how to: 

- Setup a NextJS app.
- Embed Keystone within it, and run an Admin UI you can read and write to locally.
- Create a simple Keystone schema with a `Post` [list](/apis/schema) that NextJS can build pages with. 
- Setup a secure read-only GraphQL API endpoint (and playground) that you can access in production.
- Deploy your CMS and frontend monorepo to Vercel. 🚀

?> If you need access to an Keystone’s Admin UI over the internet, or want to provide access to more than one editor, use Keystone in it’s default [standalone](/guides/modes#standalone-mode) mode.

## Setup a NextJS app

Start out by adding a basic NextJS instance empty project directory:

```bash
yarn create next-app my-project
cd my-project
```

### Add TypeScript

Keystone Next is built in TypeScript. Adding it as a `--dev` dependency will make Keystone implementation easier later on. Let‘s get that out of the way now:

```bash
yarn add --dev typescript @types/react
```
Rename the `index.js` file in your `/pages` directory to `index.tsx`. We’ll query data in Keystone using TypeScript later on.

```text
.
├── pages

> │   ├── _app.js

Author: I left `_app.js` unchanged the tut still worked. 
Just figured to leave out anything unnecessary.
If it's important from a convention standpoint happy to fold it in.

> .
> ├── pages
> │   ├── _app.js
> │   ├── api
> │   └── index.tsx // renamed

Reviewer: ```suggestion
.
└── pages
    ├── _app.js
    ├── api
    └── index.tsx // renamed
```

```

### Startup your local server

Run `yarn dev` at the root of your project. NextJS will generate a local server for you at [http://localhost:3000](http://localhost:3000).

> ![A browser showing the Home page of the default NextJS app](/assets/tutorials/next-lite/localhost-home-1.png)

Reviewer 2: We kind of skip a step here - this bit shows the default Next.js starter page, down lower (around l.142) we show our own demo homepage (unstyled; pre dynamic content) but we never show the code to paste in and replace the contents of `index.tsx` to get that to render.

I think for clarity / consistency we should add a step either before or after this one that says "replace the contents of your index.tsx page with this snippet"

No strong opinions whether that comes before an updated version of this screenshot, or after (since you're showing the default page in the animated gif lower down)

Could go below here with a "one more thing before we add Keystone... lets add the homepage for our blog, with a hard-coded index of posts for now"

## Add Keystone to your project

Now that we have the NextJS starter with static files, let‘s embed Keystone into the app to blend file-based content with content you can edit using Keystone’s intuitive Admin UI.

### Install dependencies

Add the following Keystone dependencies to the project:

```bash
yarn add @keystone-next/keystone @keystone-next/admin-ui @keystone-next/fields
```

### Update .gitignore

Add the `.keystone` directory to your `.gitignore` file. The contents of `.keystone` are dynamically generated at build time and you’ll never have to change them.

```bash

# in your .gitignore
.keystone
```

### Add Keystone to NextJS config

Add a `next.config.js` file to the project root with the following:

```tsx

// next.config.js

const { withKeystone } = require('@keystone-next/keystone/next');

module.exports = withKeystone();
```

> This is where the magic happens – the `withKeystone` constant lets NextJS encapsulate Keystone in its script runtime, but Keystone still operates independently of the NextJS frontend.✨

Reviewer 2: ```suggestion
This is where the magic happens – the `withKeystone` constant lets NextJS encapsulate Keystone in its script runtime, but Keystone still operates independently of the NextJS frontend. ✨
```

### Create a content List in Keystone config

Now add a `keystone.ts` [configuration file](/apis/config) to the project root with a simple `Post` [list](/apis/schema) type. This will give us the ability to create and edit blog records in Keystone’s Admin UI:

```tsx
// keystone.ts

import { config, list } from '@keystone-next/keystone/schema';
import { text } from '@keystone-next/fields';

const Post = list({
  fields: {
    title: text({ isRequired: true }),
    slug: text(),
    content: text(),
  },
});

export default config({
  db: { adapter: 'prisma_sqlite', url: 'file:./app.db' },
  experimental: {
    generateNextGraphqlAPI: true,
    generateNodeAPI: true,
  },
  lists: { Post },
});
```

### Update package scripts

Finally, make a small change to the `scripts` object in `package.json` to include Keystone‘s `postinstall`:

> ```json
> 	// within package.json
> 
>   "scripts": {
>     "postinstall": "keystone-next postinstall",
>     "dev": "next dev",
>     "start": "next start",
>     "build": "next build"
>   },
> ```

Reviewer: ```suggestion
```diff
  "scripts": {
+    "postinstall": "keystone-next postinstall",
    "dev": "next dev",
    "start": "next start",
    "build": "next build"
  },
```
```

Running `yarn dev` again will do the following:

- Provision a GraphQL schema based on the configuration of `keystone.ts`.
- Build a [Prisma.io](https://www.prisma.io/) schema (which Keystone uses to create tables in the database).
- Serve Keystone’s Admin UI at [http://localhost:8000](http://localhost:8000).
- Serve the NextJS frontend at [http://localhost:3000](http://localhost:3000).

> - Add a `postinstall` script that ensures everything works if we install dependencies in the future.

Reviewer 2: Yarn dev won't do this, we did it above 😛 

Also this isn't quite the purpose:

> ensures everything works if we install dependencies in the future.

It's more that it ensures the generated types / schemas are present and up to date, which can be really important for CI, but we probably don't need to get into that here

> ![Local server instances of the NExtJS frontend, and Keystone Admin UI](/assets/tutorials/next-lite/localhost-next-keystone.gif)

Reviewer 2: ```suggestion
![Local server instances of the Next.js frontend, and Keystone Admin UI](/assets/tutorials/next-lite/localhost-next-keystone.gif)
```

Next is capitalised like `Next.js`, this isn't the only place but this is the only one that also has a capital E as well 😂 

Generally need a search/replace to get the capitalisation correct across this PR

Author: good catch, will resolve!

> Go ahead and add two post entries using your Admin UI, ensuring the Slug field `uses-hyphens-and-lowercase-chars` for permalinks.

Reviewer 2: ```suggestion
Go ahead and add two post entries using your Admin UI, ensuring you only use `hyphens-and-lowercase-chars` in the slug field for permalinks.
```

Could be a great place for an info box linking out to how to set up validation for text fields, when we get to that 😉

Author: agreed. there's a great use case for a link to a specific tuts on validation, and slugifying things down the track
